src/Data/FileSystem.elm

```diff
@@ -0,0 +1,186 @@
+module Data.FileSystem exposing
+    ( FileSystem(..)
+    , FolderName
+    , filterFiles
+    , flatten
+    , mapFiles
+    , mkFileSystem
+    , reverse
+    , sortAudienceFilesAndFoldersAlphabetically
+    , sortFoldersAlphabetically
+    , sortWith
+    , toList
+    )
+
+import Data.Audience exposing (Audience)
+import Data.AudienceFolder exposing (AudienceFolder)
+import Data.List.Utility exposing (splitFilter)
+
+
+type alias FolderName =
+    String
+
+
+type FileSystem file
+    = File file
+    | Folder FolderName (List (FileSystem file))
+
+
+toList : FileSystem file -> List file
+toList tree =
+    case tree of
+        Folder name content ->
+            List.concatMap toList content
+
+        File file ->
+            [ file ]
+
+
+{-| Data.FileSystem.flatten
+
+merges all files into the root folder
+
+-}
+flatten : FileSystem file -> FileSystem file
+flatten tree =
+    case tree of
+        File _ ->
+            tree
+
+        Folder name contents ->
+            Folder name (List.map File <| toList tree)
+
+
+filterFiles : (a -> Bool) -> FileSystem a -> FileSystem a
+filterFiles check =
+    mapFolders <|
+        List.filter
+            (\ft ->
+                case ft of
+                    File a ->
+                        check a
+
+                    Folder _ _ ->
+                        True
+            )
+
+
+mapFiles : (a -> b) -> FileSystem a -> FileSystem b
+mapFiles f tree =
+    case tree of
+        File x ->
+            File (f x)
+
+        Folder name content ->
+            Folder name (List.map (mapFiles f) content)
+
+
+mapFolders : (List (FileSystem a) -> List (FileSystem a)) -> FileSystem a -> FileSystem a
+mapFolders f tree =
+    case tree of
+        File file ->
+            File file
+
+        Folder name content ->
+            Folder name <| f <| List.map (mapFolders f) content
+
+
+reverse : FileSystem a -> FileSystem a
+reverse =
+    mapFolders List.reverse
+
+
+sortWith : (FileSystem a -> FileSystem a -> Order) -> FileSystem a -> FileSystem a
+sortWith sorter =
+    mapFolders (List.sortWith sorter)
+
+
+sortFoldersAlphabetically : FileSystem a -> FileSystem a
+sortFoldersAlphabetically =
+    sortWith <|
+        \fs1 ->
+            \fs2 ->
+                case ( fs1, fs2 ) of
+                    ( Folder _ _, File _ ) ->
+                        GT
+
+                    ( File _, Folder _ _ ) ->
+                        LT
+
+                    ( Folder name1 _, Folder name2 _ ) ->
+                        compare name1 name2
+
+                    ( File _, File _ ) ->
+                        EQ
+
+
+{-| Interface with BE Audience & AudienceFolder
+
+mkFileSystem transforms BE data into FileSystem of Audiences
+
+-}
+mkFileSystem : List AudienceFolder -> List Audience -> FileSystem Audience
+mkFileSystem folders files =
```

Comment:
What is the `mk` prefix stands for? I'd either improve or explaine the choice of that prefix in the comment.

Reply:
usually, it means `make`

Reply:
I'm not even sure from where I know this convention. Is it from Haskell or Rust or Nix or something else?

Reply:
I'd go with `makeFileSystem` then.

Reply:
mk stands for make and is usually used for "dumb" constructor-like functions. It's relatively popular in Haskell (it's used all over the place in ghc compiler)

Reply:
in elm usual convention is to use `init` over `make` I would say. But that's just another cosmetic thing.